src/query/boolean_query/boolean_weight.rs

```diff
@@ -111,43 +149,88 @@ impl<TScoreCombiner: ScoreCombiner> BooleanWeight<TScoreCombiner> {
         score_combiner_fn: impl Fn() -> TComplexScoreCombiner,
     ) -> crate::Result<SpecializedScorer> {
         let mut per_occur_scorers = self.per_occur_scorers(reader, boost)?;
```

Comment:
If `minimum_number_should_match` matches the number of should scorers, we can move the scorers to the must scorers

Reply:
bump

Reply:
That would allow to remove the `FullIntersection` variant and reduce the match complexity later on